cli_tools_e2e_test/gce_image_import_export/test_suites/import/import_tests.go

```diff
@@ -83,26 +108,151 @@ var cases = []testCase{
+			os:     "windows-2019",
+		},
+		expectUEFICompatible: false,
+	},
+}
+
+func (t *testCase) run(ctx context.Context, junit *junitxml.TestCase, logger *log.Logger,
 	testProjectConfig *testconfig.Project, testType utils.CLITestType) {
+
+	t.runImportAndVerifyImage(ctx, junit, logger, testProjectConfig, testType, nil)
+}
+
+func (t *testCase) runImportAndVerifyImage(ctx context.Context, junit *junitxml.TestCase, logger *log.Logger,
+	testProjectConfig *testconfig.Project, testType utils.CLITestType,
+	verifyImageFunc func(ctx context.Context, junit *junitxml.TestCase, logger *log.Logger, testProjectConfig *testconfig.Project)) {
+
 	logger = t.createTestScopedLogger(junit, logger)
-	imageName := "e2e-test-image-import" + path.RandString(5)
-	imagePath := fmt.Sprintf("projects/%s/global/images/%s", testProjectConfig.TestProjectID, imageName)
+	t.imageName = "e2e-test-image-import" + path.RandString(5)
+	imagePath := fmt.Sprintf("projects/%s/global/images/%s", testProjectConfig.TestProjectID, t.imageName)
 
-	importLogs, err := t.runImport(junit, logger, testProjectConfig, imageName)
+	importLogs, err := t.runImport(junit, logger, testProjectConfig, t.imageName)
 
 	if t.expectedError != "" {
 		t.verifyExpectedError(junit, err, importLogs)
 	} else if err != nil {
 		t.writeImportFailed(junit, importLogs)
 	} else {
-		err = t.runPostTranslateTest(ctx, imagePath, testProjectConfig, logger)
+		if verifyImageFunc != nil {
+			verifyImageFunc(ctx, junit, logger, testProjectConfig)
+		}
+		err = t.runPostTranslateTest(ctx, imagePath, testProjectConfig, logger, strings.HasPrefix(t.os, "ubuntu"))
 		if err != nil {
 			junit.WriteFailure("Failed post translate test: %v", err)
 		}
 	}
 }
 
+func (t *inspectUEFITestCase) run(ctx context.Context, junit *junitxml.TestCase, logger *log.Logger,
+	testProjectConfig *testconfig.Project, testType utils.CLITestType) {
+
+	t.testCase.runImportAndVerifyImage(ctx, junit, logger, testProjectConfig, testType, t.verifyImage)
+	if junit.Failure != nil {
+		return
+	}
+}
+
+func (t *inspectUEFITestCase) verifyImage(ctx context.Context, junit *junitxml.TestCase, logger *log.Logger, testProjectConfig *testconfig.Project) {
+	logger.Printf("Verifying imported image...")
+	image, err := compute.CreateImageObject(ctx, testProjectConfig.TestProjectID, t.imageName)
+	if err != nil {
+		junit.WriteFailure("Image '%v' doesn't exist after import: %v", t.imageName, err)
+		logger.Printf("Image '%v' doesn't exist after import: %v", t.imageName, err)
+		return
+	}
+	logger.Printf("Image '%v' exists! Import success.", t.imageName)
+
+	guestOsFeatureForUEFICompatible := []string{"UEFI_COMPATIBLE"}
+	var expectedGuestOsFeatures, unexpectedGuestOsFeatures []string
+	if t.expectUEFICompatible {
+		expectedGuestOsFeatures = guestOsFeatureForUEFICompatible
+	} else {
+		unexpectedGuestOsFeatures = guestOsFeatureForUEFICompatible
+	}
+
+	t.verifyGuestOSFeatures(expectedGuestOsFeatures, image, junit, logger, unexpectedGuestOsFeatures)
+}
```

Comment:
I'd recommend that we didn't introspect details such as guestOSFeatures, since guestOSFeatures are an *implementation detail*. What we care about (and what our customers care about) is that their image boots, and can perform some tasks. 

I counter argument to this would be: We don't want MBR images to be tagged UEFI since they boot slower. If we care about boot times, then, the correct test would boot images and measure times, rather than indirectly look at a tag.

That said, I could understand inspecting licenses, since those generally don't have an observable side effect, but are rather present as an opaque metrics mechanism.

Reply:
In most of the cases, that is perfectly correct. This case is special, since vanadium should not be considered as completely transparent in my opinion. We know they can boot any disk with UEFI_COMPATIBLE successfully no matter whether it's really a UEFI; we also know that behavior won't be there forever (backward-compatible for BIOS disk with UEFI_COMPATIBLE can be deprecated). So, in this case, it's insufficient to only check boot-success. We also want to make sure UEFI_COMPATIBLE is marked only when needed.

Reply:
Let's meet in the middle! 
1. Undo the new interfaces, since I'm not sure there's truly a delineation between the two types (given your argument, don't we always want to very the guest os feature?)
1. Update `testCase` to have a field `expectedGuestOSFeatures` that is a list.
1. Move the assertion code outside this suite -- there's a common set of code for tests?.  I imagine a signature of `assertImageHasGuestOSFeatures(imageURI string, expectedFeatures []string)` that asserts the features are *exactly* what's passed in (nothing missing, nothing extra)
1. Call  `assertImageHasGuestOSFeatures`after `t.runPostTranslateTest`

Reply:
Hi Eric
1. To simplify the code, I removed the interface and 2 concrete impl as you suggested. I didn't apply the validation for guestOSFeatures for the existing tests. The reason is because (1) it's not what those tests want to test (2) it can be flaky. All my added tests are designed to be stable regarding the UEFI behaviors - they are based on the vmdk files that I created. Meanwhile, other tests usually use a image path, or sometimes even image family, which can be different in the future (one example: windows image was non-UEFI but now they are UEFI, cutoff date on June 2020 ). The author of the other tests don't need to verify the UEFI tag, also they don't want to carefully set the expected UEFI result when they create/modify the tests since that is not what they care about.

2. So accordingly, I updated the inspect to add a "expectedGuestOSFeatures" list as suggested, meanwhile added a "unexpectedGuestOSFeatures". The reason is in # 3

3. I moved it to the common lib as suggested so people can reuse it. The functionality is slightly different from you said: What I am verifying is whether the image contains the given guestOSFeature or not. It's not a "exact" list. It's because I don't want the test to be impacted by other features, like "IS_WINDOWS". (still kept some image-specific logic in local, since the assertGuestOSFeatures can be reused for not only image but also disk and more)

4. "Call assertImageHasGuestOSFeatures after t.runPostTranslateTest" - done similarly. It has to be before runPostTranslateTest, otherwise the image was removed by runPostTranslateTest

Reply:
> "We don't want MBR images to be tagged UEFI since they boot slower." If that's the case, then we should establish a test framework that monitors boot times, since this focuses on the observable behavior that the customer cares about.

I think we need to test whether GuestOSFeatures are set correctly by our tests. This is not a black-box for us as we know what inputs lead to which GuestOSFeatures. This is for performance reasons, but, more importantly, for correctness: UEFI_COMPATIBLE tagged MBR formatted disks might not work in the future if legacy mode is removed from the firmware.